hazelcast/src/main/java/com/hazelcast/map/impl/record/Records.java

```diff
@@ -63,4 +63,9 @@ public static Object getValueOrCachedValue(Record record, SerializationService s
         return value;
     }
 
+    private static boolean shouldCache(Record record, Object value) {
+        return (record instanceof CachedDataRecord || record instanceof CachedDataRecordWithStats)
```

Comment:
I would simplify this a bit. e.g.

```
if(!((record instanceof CachedDataRecord || record instanceof CachedDataRecordWithStats))
   return false;

if(....
```

The current implementation is painful to read.